File: game.py
import time
from collections import defaultdict

from planetwars import Fleet, Planet
from planetwars.internal import load_all_maps
from planetwars.utils import count_ships, partition

def _neutral_player(turn, pid, planets, fleets):
    return []

class PlanetWars:

    ais = {}
    maps = load_all_maps()

    def __init__(self, players, map_name, turns_per_second=2):
        if len(players) < 2:
            raise Exception("A game requires at least two players.")
        self.player_names = players
        self.players = [_neutral_player] + [PlanetWars.ais[player] for player in players]
        self.map_name = map_name
        planets, fleets = PlanetWars.maps[map_name]
        self.planets = [Planet(*planet) for planet in planets]
        self.fleets = [Fleet(*fleet) for fleet in fleets]
        self.views = []
        self.turns_per_second = turns_per_second
        self.turn_duration = 1.0 / turns_per_second
        self.turn = 0

    def add_view(self, view):
        self.views.append(view)

    def freeze(self):
        planets = tuple(planet.freeze() for planet in self.planets)
        fleets = tuple(fleet.freeze() for fleet in self.fleets)
        return planets, fleets

    def play(self):
        planets, fleets = self.freeze()
        for view in self.views:
            view.initialize(self.turns_per_second, self.planets, self.map_name, self.player_names)
            view.update(planets, fleets)
        next_turn = time.time() + self.turn_duration
        winner = -1
        while winner < 0:
            # Wait until time has passed
            now = time.time()
            if now < next_turn:
                time.sleep(next_turn - now)
            next_turn += self.turn_duration
            # Do the turn
            self.do_turn()
            # Update views
            planets, fleets = self.freeze()
            for view in self.views:
                view.update(planets, fleets)
            # Check for end game.
            winner, ship_counts = self.gameover()
        for view in self.views:
            view.game_over(winner, ship_counts)

    def do_turn(self):
        """Performs a single turn of the game."""

        # Get orders
        planets, fleets = self.freeze()
        player_orders = [player(self.turn, i, planets, fleets) for i, player in enumerate(self.players)]
        self.turn += 1

        # Departure
        for player, orders in enumerate(player_orders):
            for order in orders:
                self.issue_order(player, order)

        # Advancement
        for planet in self.planets:
            planet.generate_ships()
        for fleet in self.fleets:
            fleet.advance()

        # Arrival
        arrived_fleets, self.fleets = partition(lambda fleet: fleet.has_arrived(), self.fleets)
        for planet in self.planets:
            planet.battle([fleet for fleet in arrived_fleets if fleet.destination == planet])

    def issue_order(self, player, order):
        if order.source.owner != player:
            raise Exception("Player %d issued an order from enemy planet %d." % (player, order.source.id))
        source = self.planets[order.source.id]
        ships = int(min(order.ships, source.ships))
        if ships > 0:
            destination = self.planets[order.destination.id]
            source.ships -= ships
            self.fleets.append(Fleet(player, ships, source, destination))

    def gameover(self):
        players = range(1, len(self.players))
        living = list(filter(self.is_alive, players))
        if len(living) == 1:
            return living[0], count_ships(self.planets, self.fleets)
        elif self.turn >= 200:
            ship_counts = count_ships(self.planets, self.fleets)
            ship_counts = [(p, s) for p, s in ship_counts if p > 0]
            winner = 0 if ship_counts[0][1] == ship_counts[1][1] else ship_counts[0][0]
            return winner, ship_counts
        else:
            return -1, []

    def is_alive(self, player):
        for planet in self.planets:
            if planet.owner == player:
                return True
        return False

File: game.js
const cloneDeep = require('lodash/cloneDeep');

const { Fleet } = require('./dataTypes');
const { countShips } = require('./utils');
const { loadAllAIs, loadAllMaps } = require('./internal');

const AIS = loadAllAIs();
const MAPS = loadAllMaps();

class NeutralPlayer {
  play() {
    return [];
  }
}

function makeAI(name) {
  return new AIS[name]();
}

class PlanetWars {
  constructor(players, mapName, turnsPerSecond=2) {
    if (players.length < 2) {
      throw new Error('A game requires at least two players.');
    }
    this.playerNames = players;
    this.players = [new NeutralPlayer()].concat(players.map(makeAI));
    this.mapName = mapName;
    let [planets, fleets] = MAPS[mapName];
    this.planets = cloneDeep(planets);
    this.fleets = cloneDeep(fleets);
    this.views = [];
    this.turnsPerSecond = turnsPerSecond;
    this.turnDuration = 1000 / turnsPerSecond;
    this.turn = 0;
  }

  addView(view) {
    this.views.push(view);
  }

  freeze() {
    return cloneDeep([this.planets, this.fleets]);
  }

  play() {
    for (let view of this.views) {
      view.initialize(
        this.turnsPerSecond, this.planets, this.mapName, this.playerNames);
      view.update(this.planets, this.fleets);
    }
    this.doPlay();
  }

  doPlay() {
    let nextTurn = Date.now() + this.turnDuration;
    this.doTurn();
    for (let view of this.views) {
      view.update(this.planets, this.fleets);
    }
    let [winner, shipCounts] = this.isGameOver();
    if (winner < 0) {
      setTimeout(() => this.doPlay(), nextTurn - Date.now());
    } else {
      for (let view of this.views) {
        view.gameOver(winner, shipCounts);
      }
    }
  }

  doTurn() {
    const play = (player, i) => player.play(
      this.turn, i, cloneDeep(this.planets), cloneDeep(this.fleets));
    // Get orders.
    let playerOrders = this.players.map(play);
    this.turn += 1;

    // Send fleets.
    //for (let [player, orders] of playerOrders) {
    playerOrders.forEach((orders, player) => {
      for (let order of orders) {
        this.issueOrder(player, order);
      }
    });

    // Make ships.
    for (let planet of this.planets) {
      planet.generateShips();
    }

    // Advance fleets.
    for (let fleet of this.fleets) {
      fleet.advance();
    }

    let arrivedFleets = this.fleets.filter((fleet) => fleet.hasArrived());
    this.fleets = this.fleets.filter((fleet) => !fleet.hasArrived());
    for (let planet of this.planets) {
      planet.battle(
        arrivedFleets.filter((fleet) => fleet.destination === planet));
    }
  }

  issueOrder(player, order) {
    if (order.source.owner !== player) {
      throw new Error(`Player ${player} issued an order from`
        + `enemy planet ${order.source.id}.`);
    }
    let source = this.planets[order.source.id];
    // Don't let a player send more ships than they have.
    let ships = Math.floor(Math.min(order.ships, source.ships));
    if (ships > 0) {
      let destination = this.planets[order.destination.id];
      source.ships -= ships;
      this.fleets.push(new Fleet(player, ships, source, destination));
    }
  }

  isGameOver() {
    let playerIds = this.players.map((p, i) => i).slice(1);
    let living = playerIds.filter((id) => this.isAlive(id));
    if (living.length === 1) {
      return [living[0], countShips(this.planets, this.fleets)];
    } else if (this.turn >= 200) {
      let shipCounts = countShips(this.planets, this.fleets);
      shipCounts = shipCounts.filter(([p]) => p > 0);
      let winner = shipCounts[0][1] === shipCounts[1][1] ? 0 : shipCounts[0][0];
      return [winner, shipCounts];
    }
    return [-1, []];
  }

  isAlive(playerId) {
    for (let planet of this.planets) {
      if (planet.owner === playerId) {
        return true;
      }
    }
    return false;
  }
}

module.exports = { PlanetWars , AIS, MAPS };

